src/main/io/osd.c

```diff
@@ -425,19 +430,31 @@ static void osdDrawSingleElement(uint8_t item)
 #endif // GPS
 
     case OSD_COMPASS_BAR:
-        {
-            int16_t h = DECIDEGREES_TO_DEGREES(attitude.values.yaw);
-
-            h = osdGetHeadingIntoDiscreteDirections(h, 16);
+    {
+        int16_t h = DECIDEGREES_TO_DEGREES(attitude.values.yaw);
+#ifdef USE_GPS
+		// If MAG_Heading is not available, yaw starts with zero in a random direction. With GPS course over ground the flight direction is roughly displayed instead
+		if (STATE(GPS_FIX) && STATE(GPS_FIX_HOME) && !sensors(SENSOR_MAG | SENSOR_RANGEFINDER)) {
+			h = DECIDEGREES_TO_DEGREES(gpsSol.groundCourse);
+		}
+#endif
+        h = osdGetHeadingIntoDiscreteDirections(h, 16);
 
-            memcpy(buff, compassBar + h, 9);
-            buff[9]=0;
-            break;
-        }
+        memcpy(buff, compassBar + h, 9);
+        buff[9]=0;
+        break;
+    }
 
     case OSD_ALTITUDE:
         {
+#ifdef USE_GPS
+			// If BARO is not available, the estimated altitude is zero. Using GPS altitude an absolute height can be displayed instead.
+			if (STATE(GPS_FIX) && !sensors(SENSOR_BARO | SENSOR_RANGEFINDER)) {
+				osdFormatAltitudeString(buff, gpsSol.llh.alt*10, true); // if GPS altitude is in 0.1m units
```

Comment:
Again, this should really be done in the altitude estimation code (even if it just replaces the estimate with the gps altitude if available). Having OSD display one value for altitude, while everything else is using a different value is confusing and hard to maintain..